notes/conds.md

```diff
@@ -0,0 +1,83 @@
+| ------------ | --------------------------------- | ------------------------------------------ | ---------------------------------------------- |
+| `/${string}` |`[] -> Any`                        | Selects value using JSON Pointer            | `["/foo/0"]`                                  |
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `some`       |`[Expr<Any[]>, Expr<Bool>] -> Bool`| Collection predicate                        | `["some", ["/to"], ["==", [], "hi@web.mail"]]`|
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+     ["$ARGS", "foo..bar.[].baz", "?x"],
+     [["some", "match"], "?x", "*@example.com"],
+      
+     ["$ARGS", "foo.quux", "?y"],
+     ["?y", ">", 42],
+     ["?x", "<", "?y"]
+  ]
+}
+```
```

Comment:
I think representing references as tuples would be both unambiguous and very flexible, this is how this example would look through this frame

```suggestion
```js
{
  "iss": "did:key:alice",
  // ...
  "policy": [
     [[], "foo..bar.[].baz", ["x"]],
     [["x"], { match: {} }, "*@example.com"],
     [[], "foo.quux", ["y"]],
     [["y"], {">":{}}, 42],
     [["x"], {"<":{}}, ["y"]]
  ]
}
```
```

What did I change:

1. Refs are now arrays
2. Empty array is reference to the context
3. Operators are now objects. This also means we could add modifiers e.g. `{match:{every:true}}` I assume that they are all implicitly some.

Reply:
And yes one may write something like

```js
[["y"], {">":{}, "<": {}}, 42],
```

Which would be equivalent of (and never true)

```js
[["y"], {">":{}}, 42],
[["y"], {"<": {}}, 42]
```

Reply:
On a separate note I really like that `every` becomes a modifier of the operator (relation) because it really is a modifier.

Reply:
If you really don't like objects and arrays I would suggest something like this as an alternative

1. Selectors start with leading `.` making them distinct from operators (implied relations)
1. Relation cardinality can be changed using `!` suffix, that is a modifier that switches from `some` to `every`.
1. `?` variable is bound to the context

```suggestion
{
  "iss": "did:key:alice",
  // ...
  "policy": [
     ["?", ".foo..bar.[].baz", "?x"],
     ["?x", "match!", "*@example.com"],
      
     ["?", ".foo.quux", "?y"],
     ["?y", ">", 42],
     ["?x", "<", "?y"]
  ]
}
```

Reply:
```
      [[], "foo..bar.[].baz", ["x"]],
	     [["x"], {match:{}}, "*@example.com"],
	     [[], "foo.quux", ["y"]],
	     [["y"], {">":{}}, 42],
	     [["x"], {"<":{}}, ["y"]]
```

I find this difficult to read 🤔 Also I was hoping to be able to substitute the literals equivalent to a variable and have things continue to work (e..g when debugging). Was this maybe before we talked about array/object literals?

Also FWIW I'm poking around the Datomic docs, and they also use `$` and `?` to distinguish kernel vs userspace variables.

<img width="461" alt="Screenshot 2024-02-24 at 17 03 06" src="https://github.com/ucan-wg/delegation/assets/1052016/60c095fc-8587-40bf-a9d5-47e5c2ea93e2">

Reply:
> Selectors start with leading . making them distinct from operators (implied relations)

Yep makes sense 👍

Reply:
> ? variable is bound to the context

Specifically to `args`, I guess? Sure, I can live with this. It does limit how we can extend it later, though (e.g. let's say that we change our minds on `exp`)

Reply:
> Relation cardinality can be changed using ! suffix, that is a modifier that switches from some to every.

That's pretty interesting! It does mean that we need `!` versions of everything. I _think_ we can do this with `some`/`every` operators (see [that ](https://github.com/ucan-wg/delegation/pull/3/files#r1501720168)

Reply:
> That's pretty interesting! It does mean that we need `!` versions of everything. I _think_ we can do this with `some`/`every` operators ([see this](https://github.com/ucan-wg/delegation/pull/3/files#r1501720168))

Not on logic operators I think